### DISCUSSION TITLE
Spark 3.* bundles transitively depend on Spark 2.4

### DISCUSSION THREAD

**Discussion by yuribogomolov** [AbsaOSS/spline-spark-agent#516] [Q&A]
The `agent-core` package version is agnostic to the Spark version and it depends on Spark 2.4. In result all Spark 3.* bundles have transitive dependencies on Spark 2.4. See the dependency tree for `spark-3.2-spline-agent-bundle_2.12` below:
````
[INFO] +- za.co.absa.spline.agent.spark:spark-3.2-spline-agent-bundle_2.12:jar:0.7.8:compile
[INFO] |  +- za.co.absa.spline.agent.spark:agent-core_2.12:jar:0.7.8:compile
[INFO] |  |  +- org.apache.spark:spark-core_2.12:jar:2.4.2:compile
```

**Comment by cerveada** [comment]
When the bundles are created profiles are used. So to find out actual dependencies maven should be called with the profile. 
```
mvn dependency:tree -Pscala-2.12,spark-3.0
```

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ spark-3.2-spline-agent-bundle_2.12 ---
[INFO] za.co.absa.spline.agent.spark:spark-3.2-spline-agent-bundle_2.12:jar:1.0.0-SNAPSHOT
[INFO] +- za.co.absa.spline.agent.spark:agent-core_2.12:jar:1.0.0-SNAPSHOT:compile
[INFO] |  +- org.scala-lang:scala-compiler:jar:2.12.15:provided
[INFO] |  +- za.co.absa.commons:commons_2.12:jar:1.2.0:compile
[INFO] |  +- org.apache.spark:spark-core_2.12:jar:2.4.8:provided
```

From this you can see the dependency is provided and should not have any influence further down the line.

But the bundle itself is **not supposed to be used as maven dependency** anyway. It is an artifact ready to use with the specific Spark and Scala version. So it doesn't really matter what it declares dependency on.

https://github.com/AbsaOSS/spline-spark-agent#selecting-artifact

**Comment by yuribogomolov** [comment]
Thank you for the fast and detailed response!

Maven packages that contain a spark version `X` in their version usually depend on the Spark version `X`. E.g. [spark-testing-base_2.12/3.2.1_1.2.0](https://mvnrepository.com/artifact/com.holdenkarau/spark-testing-base_2.12/3.2.1_1.2.0) depends on Spark 3.2.1, or [spark-cassandra-connector_2.12/3.2.0](https://mvnrepository.com/artifact/com.datastax.spark/spark-cassandra-connector_2.12/3.2.0) depends on Spark 3.2. That's why I was confused that `spark-3.2-spline-agent-bundle_2.12` transitively depends on spark 2.4.

Looks like `agent-core` is the recommended way to define Spline agent dependencies for users with custom layers (like `PostProcessingFilter` or `Plugin`) who need to rebuild the jar?

**Reply by cerveada** [reply]
> Looks like agent-core is the recommended way to define Spline agent dependencies for users with custom layers (like PostProcessingFilter or Plugin) who need to rebuild the jar?

Yes, the bundles are what is called fat-jars (or uber-jars). In short they pack inside their dependencies. Generally you should not declare dependencies on fat-jars and it's true for Spline bundles as well.

For PostProcessingFilter or Plugin you just need to create new empty project and declare the dependency on `agent-core` then you can write your filter/plugin and package it as jar. Later you provide that jar using `spark-submit --jars`. Then you just tell Spline to load your filter/plugin via configuration.

Using this process you should not need to do any changes to spline code itself.
